src/matchengine/persist.rs

```diff
@@ -240,44 +241,79 @@ pub async fn init_from_db(conn: &mut ConnectionType, controller: &mut Controller
     Ok(())
 }
 
+const DUMPING_SET_LIMIT: usize = 10000;
+
+fn collect_n<T: std::iter::Iterator>(iter: &mut T, n: usize, mut record: Vec<T::Item>) -> Vec<T::Item> {
+    if record.len() >= n {
+        return record;
+    }
+
+    for i in iter {
+        record.push(i);
+        if record.len() >= n {
+            return record;
+        }
+    }
+
+    record
+}
+
+async fn dump_records<Q, T>(mut iter: T, n: usize, conn: &mut ConnectionType) -> anyhow::Result<usize>
+where
+    Q: Clone + TableSchemas,
+    Q: for<'r> SqlxAction<'r, InsertTable, DbType>,
+    T: std::iter::Iterator<Item = Q>,
+{
+    let mut records: Vec<Q> = collect_n(&mut iter, n, Vec::new());
```

Comment:
seems we can make use of https://doc.rust-lang.org/std/vec/struct.Vec.html#method.truncate

and no need to implement `collect_n`

Reply:
for example,

```
let all : Vec<Q> = iter.collect();
let records = all.truncate(n);
```

not tested yet, but the idea is there

Reply:
Sorry I have made a mistake here .... I should also call collect_n inside the while loop to read all elements inside the iterator, and keep filling the records to a maximum size (so batch insert achieve maximum efficiency)

The code currently we have merged is oblivious wrong: it drop the elements which should be persisted if the total size larger than DUMPING_SET_LIMIT.